File: 프린터.cpp
#include <string>
#include <vector>

using namespace std;

//아이디어: 배열을 순회하며 더 큰값이 있으면 현재위치 배열을 맨 뒤에 추가
//			더 큰 값이 없으면 answer에 1을 더함
//			만약 현재위치가 찾고자 하는 값이면 answer를 반환
int solution(vector<int> priorities, int location) {
	int answer = 0;
	
	//priorities 배열 순회 2중 반복문
	for (int i = 0; i < priorities.size(); i++) {
		for (int j = i; j < priorities.size(); j++) {

			//현재 인덱스보다 더 큰 값이 있으면 현재 배열 값을 배열 뒤에 추가
			if (priorities[i] < priorities[j]) {
				priorities.push_back(priorities[i]);

				//현재 인덱스가 찾고자 하는 위치였다면 인덱스 갱신
				if (i == location)
					location = priorities.size() - 1;

				break;
			}

			//더 큰값이 없다면 answer에 1을 추가하고 
			//찾고자 하는 위치였다면 answer를 반환
			if (j + 1 == priorities.size()) {
				answer++;
				if (i == location)
					return answer;
			}
		}
	}
}

File: STACK/프린터.py
"""
    Queue를 사용해 풀었습니다.
    우선순위가 낮을경우 인덱스 값이 계속 변하기 때문에
    인덱스 위치를 저장할 idxQ 를 생성해 value가 뒤로 밀리면
    인덱스 값도 뒤로 밀리게 설계했습니다.
"""

from collections import deque


def solution(priorities, location):
    answer = 0

    queue = deque(priorities)
    idxQ = deque([i for i in range(len(queue))])

    while len(queue) > 1:
        now = queue.popleft()
        idx = idxQ.popleft()

        if now >= max(queue):
            answer += 1
            if idx is location:
                return answer
        else:
            queue.append(now)
            idxQ.append(idx)

    return answer + 1